storage/redis_adapter.go

```diff
@@ -24,35 +38,63 @@ func NewRedisStorageAdapter(bufferSize int) (*redisAdapter, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &redisAdapter{
+	rsa := &redisAdapter{
 		bufferSize: bufferSize,
 		redisClient: r.NewClient(&r.Options{
 			Addr:     fmt.Sprintf("%s:%d", cfg.RedisHost, cfg.RedisPort),
 			Password: cfg.RedisPassword, // "" == no password
 			DB:       int64(cfg.RedisDB),
 		}),
-	}, nil
+		messageChannel: make(chan *message),
+	}
+	go rsa.start()
+	return rsa, nil
 }
 
-// Write adds a log message to to an app-specific list in redis using ring-buffer-like semantics
-func (a *redisAdapter) Write(app string, message string) error {
-	// Note: Deliberately NOT using MULTI / transactions here since in this implementation of the
-	// redis client, MULTI is not safe for concurrent use by multiple goroutines. It's been advised
-	// by the authors of the gopkg.in/redis.v3 package to just use pipelining when possible...
-	// and here that is technically possible. In the WORST case scenario, not having transactions
-	// means we may momentarily have more than the desired number of log entries in the list /
-	// buffer, but an LTRIM will eventually correct that, bringing the list / buffer back down to
-	// its desired max size.
-	pipeline := a.redisClient.Pipeline()
-	if err := pipeline.RPush(app, message).Err(); err != nil {
-		return err
-	}
-	if err := pipeline.LTrim(app, int64(-1*a.bufferSize), -1).Err(); err != nil {
-		return err
+func (a *redisAdapter) start() {
+	messageCount, pipeline, timeoutTicker, queuedApps := a.resetMessageQueue()
+	for {
+		select {
+		case message := <-a.messageChannel:
+			if err := pipeline.RPush(message.app, message.messageBody).Err(); err == nil {
+				(*queuedApps)[message.app] = true
+				messageCount++
+				if messageCount == 50 {
+					messageCount, pipeline, timeoutTicker, queuedApps = a.execPipeline(pipeline, queuedApps)
+				}
+			} else {
+				log.Printf("Error adding rpush to %s to the pipeline: %s", message.app, err)
+			}
+		case <-timeoutTicker.C:
+			messageCount, pipeline, timeoutTicker, queuedApps = a.execPipeline(pipeline, queuedApps)
+		}
 	}
-	if _, err := pipeline.Exec(); err != nil {
-		return err
+}
+
+func (a *redisAdapter) execPipeline(pipeline *r.Pipeline, queuedApps *map[string]bool) (int, *r.Pipeline, *time.Ticker, *map[string]bool) {
+	for app := range *queuedApps {
+		if err := pipeline.LTrim(app, int64(-1*a.bufferSize), -1).Err(); err != nil {
+			log.Printf("Error adding ltrim of %s to the pipeline: %s", app, err)
+		}
 	}
+	go func() {
+		if _, err := pipeline.Exec(); err != nil {
+			log.Printf("Error executing pipeline: %s", err)
+		}
+	}()
+	return a.resetMessageQueue()
+}
+
+func (a *redisAdapter) resetMessageQueue() (int, *r.Pipeline, *time.Ticker, *map[string]bool) {
+	return 0,
+		a.redisClient.Pipeline(),
+		time.NewTicker(time.Second),
+		&map[string]bool{}
```

Comment:
Maps are implicitly references? I was not aware. Thanks! This will also probably get fixed naturally when I encapsulate more of the message queuing mechanism in its own type.